src/cursor/common.rs

```diff
@@ -214,3 +249,43 @@ pub(crate) struct CursorInformation {
     pub(crate) batch_size: Option<u32>,
     pub(crate) max_time: Option<Duration>,
 }
+
+#[derive(Clone, Debug)]
+pub(crate) struct PinnedConnection {
+    connection: Option<Arc<Mutex<Connection>>>,
```

Comment:
It's hypothetically possible that this doesn't need to be an `Arc<Mutex<_>>`, but could instead use a similar trick to the `Session` associated type in `GetMoreProviderResult`.  However, couldn't get that to line up well after a day or so of work - particular challenges were the manual polling and state management in the cursor lifecycle, the drop impls for the cursors, and not having lifetimes leak out into the public cursor api - so I went with this solution.

There is a runtime cost incurred by the reference counting and locking, but it's unlikely to be a significant issue in practice because the cursors buffer values returned, so there's not going to be a hot loop involving those operations.

Reply:
Yeah I agree this is probably fine, and it's simple which is really nice considering the complexity of the existing session stuff. The runtime cost from the locking at least will be insignificant I think since cursors themselves can't be used concurrently.